x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.test.ts

```diff
@@ -0,0 +1,1006 @@
+        ...newToken,
+        access_all_engines: false,
+        engines: ['someEngine', 'anotherEngine'],
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.setAccessAllEngines(true);
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        activeApiToken: expect.any(Object),
+      });
+    });
+  });
+
+  describe('onApiKeyDelete', () => {
+    describe('apiTokens', () => {
+      it('should remove specified token from apiTokens if name matches', () => {
+        mount({
+          apiTokens: [newToken],
+        });
+
+        CredentialsLogic.actions.onApiKeyDelete(newToken.name);
+        expect(CredentialsLogic.values.apiTokens).toEqual([]);
+      });
+
+      it('should not remove specified token from apiTokens if name does not match', () => {
+        mount({
+          apiTokens: [newToken],
+        });
+
+        CredentialsLogic.actions.onApiKeyDelete('foo');
+        expect(CredentialsLogic.values.apiTokens).toEqual([newToken]);
+      });
+    });
+
+    it('should not change any other values', () => {
+      mount();
+      CredentialsLogic.actions.onApiKeyDelete('foo');
+      expect(CredentialsLogic.values).toEqual({
+        ...DEFAULT_VALUES,
+        apiTokens: expect.any(Array),
+      });
+    });
+  });
+
+  describe('onApiTokenCreateSuccess', () => {
+    describe('apiTokens', () => {
+      const existingToken = {
+        name: 'some_token',
+        type: PRIVATE,
+      };
+
+      it('should add the provided token to the apiTokens list', () => {
+        mount({
+          apiTokens: [existingToken],
+        });
+
+        CredentialsLogic.actions.onApiTokenCreateSuccess(newToken);
+        expect(CredentialsLogic.values.apiTokens).toEqual([existingToken, newToken]);
+      });
+    });
+
+    describe('activeApiToken', () => {
+      // TODO It is weird that methods like this update activeApiToken but not activeApiTokenIsExisting...
+      it('should reset to the default value, which effectively clears out the current form', () => {
+        mount({
+          activeApiToken: newToken,
+        });
+
+        CredentialsLogic.actions.onApiTokenCreateSuccess(newToken);
+        expect(CredentialsLogic.values.activeApiToken).toEqual(DEFAULT_VALUES.activeApiToken);
+      });
+    });
+
+    describe('activeApiTokenRawName', () => {
+      it('should reset to the default value, which effectively clears out the current form', () => {
+        mount({
+          activeApiTokenRawName: 'foo',
+        });
+
+        CredentialsLogic.actions.onApiTokenCreateSuccess(newToken);
+        expect(CredentialsLogic.values.activeApiTokenRawName).toEqual(
+          DEFAULT_VALUES.activeApiTokenRawName
+        );
+      });
+    });
+
+    describe('showCredentialsForm', () => {
+      it('should reset to the default value, which closes the credentials form', () => {
+        mount({
+          showCredentialsForm: true,
```

Comment:
On nearly every test, I mount the logic with an explicit value for the value under test. This is to ensure that the value is _actually_ changing (to avoid a false positive test if the default value matches what the test is setting it to), and also makes it more readable, as a reader doesn't need to hop up to the top of the file to see what the original value of `showCredentialsForm` was.

Reply:
Just wanted to say that reading through the tests, I really really liked this. 👍 👍 👍  It did indeed make it super easy to compare initial state vs end state. Awesome stuff!!!